File: articles-api-typescript/src/entities/Article.ts
import "reflect-metadata";
import { Entity, PrimaryColumn, Column, ManyToMany, JoinTable } from "typeorm";
import { v4 as uuidv4 } from "uuid";
import { Tag } from "./Tag";

/**
 * Converts a title to a URL-friendly slug.
 * Mirrors the Java Article.toSlug() regex:
 *   title.toLowerCase().replaceAll("[\\&|[\\uFE30-\\uFFA0]|\\'|\\"|\\s\\?\\,\\.]+", "-")
 */
export function toSlug(title: string): string {
  return title
    .toLowerCase()
    .replace(/[&|\uFE30-\uFFA0'"\s?,.[\]]+/g, "-");
}

@Entity("articles")
export class Article {
  @PrimaryColumn("varchar")
  id: string;

  @Column("varchar")
  slug: string;

  @Column("varchar")
  title: string;

  @Column("varchar")
  description: string;

  @Column("text")
  body: string;

  @Column("varchar")
  userId: string;

  @Column("datetime")
  createdAt: Date;

  @Column("datetime")
  updatedAt: Date;

  @ManyToMany(() => Tag, { cascade: true, eager: true })
  @JoinTable({
    name: "article_tags",
    joinColumn: { name: "articleId", referencedColumnName: "id" },
    inverseJoinColumn: { name: "tagId", referencedColumnName: "id" },
  })
  tags!: Tag[];

  constructor(
    title: string,
    description: string,
    body: string,
    tagList: string[],
    userId: string
  ) {
    this.id = uuidv4();
    this.title = title ?? "";
    this.slug = toSlug(this.title);
    this.description = description ?? "";
    this.body = body ?? "";
    this.userId = userId ?? "";
    const now = new Date();
    this.createdAt = now;
    this.updatedAt = now;
  }

  update(title?: string, description?: string, body?: string): void {
    if (title && title.length > 0) {
      this.title = title;
      this.slug = toSlug(title);
      this.updatedAt = new Date();
    }
    if (description && description.length > 0) {
      this.description = description;
      this.updatedAt = new Date();
    }
    if (body && body.length > 0) {
      this.body = body;
      this.updatedAt = new Date();
    }
  }
}

File: article-service/src/main/java/io/spring/articleservice/model/Article.java
package io.spring.articleservice.model;

import jakarta.persistence.CascadeType;
import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.Id;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.JoinTable;
import jakarta.persistence.ManyToMany;
import jakarta.persistence.Table;
import java.time.Instant;
import java.util.HashSet;
import java.util.Set;
import java.util.UUID;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

@Entity
@Table(name = "articles")
@Getter
@Setter
@NoArgsConstructor
public class Article {

  @Id
  @Column(length = 255)
  private String id;

  @Column(nullable = false, unique = true, length = 255)
  private String slug;

  @Column(nullable = false, length = 255)
  private String title;

  @Column(columnDefinition = "TEXT")
  private String description;

  @Column(columnDefinition = "TEXT")
  private String body;

  @Column(name = "user_id", nullable = false, length = 255)
  private String userId;

  @Column(name = "created_at", nullable = false)
  private Instant createdAt;

  @Column(name = "updated_at", nullable = false)
  private Instant updatedAt;

  @ManyToMany(cascade = {CascadeType.PERSIST, CascadeType.MERGE})
  @JoinTable(
      name = "article_tags",
      joinColumns = @JoinColumn(name = "article_id"),
      inverseJoinColumns = @JoinColumn(name = "tag_id"))
  private Set<Tag> tags = new HashSet<>();

  public Article(String title, String description, String body, String userId) {
    this.id = UUID.randomUUID().toString();
    this.slug = toSlug(title);
    this.title = title;
    this.description = description;
    this.body = body;
    this.userId = userId;
    this.createdAt = Instant.now();
    this.updatedAt = Instant.now();
  }

  public void update(String title, String description, String body) {
    if (title != null && !title.isEmpty()) {
      this.title = title;
      this.slug = toSlug(title);
      this.updatedAt = Instant.now();
    }
    if (description != null && !description.isEmpty()) {
      this.description = description;
      this.updatedAt = Instant.now();
    }
    if (body != null && !body.isEmpty()) {
      this.body = body;
      this.updatedAt = Instant.now();
    }
  }

  public static String toSlug(String title) {
    return title.toLowerCase().replaceAll("[\\&|[\\uFE30-\\uFFA0]|\\'|\\\"\\s\\?,\\.]+", "-");
  }
}
